aspen/http/resource.py

```diff
@@ -29,25 +32,66 @@ class Dynamic(Simplate):
     """Model a dynamic HTTP resource using simplates.
 
        Make .request_processor available as it has been historically.
-
-       Figure out which accept header to use.
-
     """
 
-    def __init__(self, request_processor, fs, raw, default_media_type):
+    def __init__(self, request_processor, fs, raw, fs_media_type):
         self.request_processor = request_processor
+        self.fs_media_type = fs_media_type
         defaults = request_processor.simplate_defaults
+        default_media_type = fs_media_type or request_processor.media_type_default
         super(Dynamic, self).__init__(defaults, fs, raw, default_media_type)
 
-
     def render(self, state):
+        """Render the resource with the given state as context, return Output.
+
+        Before rendering we need to determine what type of content we're going
+        to send back, by trying to find a match between the media types the
+        client wants and the ones provided by the resource.
+
+        The two sources for what the client wants are the extension in the
+        request URL, and the Accept header. If the former fails to match we
+        raise NotFound (404), if the latter fails we raise NegotiationFailure
+        (406).
+
+        Note that we don't always respect the `Accept` header (the spec says
+        we can ignore it: <https://tools.ietf.org/html/rfc7231#section-5.3.2>).
+        """
+        available = self.available_types
+        choice = 0  # default to first available type
+        # When there is an extension in the URI path, the dispatcher gives us the
+        # corresponding media type (or an empty string if unknown)
         accept = dispatch_accept = state['dispatch_result'].extra.get('accept')
-        if accept is None:
-            accept = state.get('accept_header')
-        try:
-            return super(Dynamic, self).render(accept, state)
-        except NegotiationFailure:
-            if dispatch_accept is not None:
-                # e.g. client requested `/foo.json` but `/foo.spt` has no JSON page
+
+        if dispatch_accept is not None:
+            # If the extension is unknown, raise NotFound
```

Comment:
I'm thinking of law browsers such as http://www.ecode360.com/AM2259 or http://www.ecfr.gov/cgi-bin/text-idx?SID=a457db49113d5f8ca45f5252dd7dced0&mc=true&node=se31.3.1010_1100&rgn=div8. You might have on the filesystem:

`www/%section.spt`

Maybe you are parsing `request.path['section']` further, or maybe you're passing it through to a database (`SELECT * FROM sections WHERE section=%s`).

I agree it's an edge case, and I'm okay with merging this PR as-is and revisiting if/when we hit this edge case in the future.